#' Execute \code{apt-copynumber-cyto}.
#' 
#' Method to exceute the command-line call created with \code{APTparam} to
#' the command-line tool \code{apt-copynumber-cyto}.
#' 
#' @export run
setMethod( "run", 
  signature("APTCopyNumberParam"),
  function(object) {
    tool <- system.file(paste0("exec", 
        .Platform$file.sep, object@tool), package="affy2sv")

    if(object@os == "GNU/Linux") {
      if(object@level == "custom") {
        object@call <- paste0(tool, " ", object@params)
      } else {
        object@call <- paste0(tool, " -v 4",
          " --cyto2 false",
          " --doDualNormalization true",
          " --keep-intermediate-data false",
          " --run-geno-qc true", 
          " --snp-qc-use-contrast true",
          " --force true",
          " --adapter-type-normalization false",
          " --text-output true",
          " --cnchp-output false",
          " --cychp-output true",
          " --set-analysis-name cyhd",
          " --analysis loh-segment",
          " --analysis cn-neutral-loh",
          " --male-gender-ratio-cutoff 1.5",
          " --female-gender-ratio-cutoff 0.9",
          " --xx-cutoff 0.61",
          " --xx-cutoff-high 0.95",
          " --y-cutoff 0.58",
          " --local-gc-background-intensity-adjustment-method none",
          " --image-correction-intensity-adjustment-method none",
          " --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=6.wave-smooth=true",
          " --cn-calibrate-parameters calibrated-log2ratios.alpha-cn-calibrate=0.564278.alpha-X-cn-calibrate=0.619453.alpha-Y-cn-calibrate=0.494620.beta-cn-calibrate=1.beta-X-cn-calibrate=1.beta-Y-cn-calibrate=1",
          " --analysis genotype --analysis log2-ratio.gc-correction=false.median-autosome-median-normalization=true.median-smooth-marker-count=5 --log2ratio-adjustment-method log2ratio-adjustment-method-high-pass-filter.use=true",
          " --analysis allelic-difference-CytoScan.outlier-trim=3.0.step=20.window=100.point-count=128.bandwidth=0.25.cutoff=0.05.clean-threshold=0.35.symmetry=true ",
          " --analysis kernel-smooth.sigma_span=50",
          " --analysis cn-cyto2.",
          "hmmCN_state=\\'0,1,2,3,4\\'.",
          "hmmCN_mu=\\'-2,-0.45,0,0.3,0.51\\'.",
          "hmmCN_sigma=\\'0.35,0.35,0.25,0.25,0.25\\'.",
          "hmmCN_state-X=\\'0,1,2,3,4\\'.",
          "hmmCN_mu-X=\\'-2,-0.47,0,0.33,0.53\\'.",
          "hmmCN_sigma-X=\\'0.35,0.35,0.25,0.25,0.25\\'.",
          "hmmCN_state-Y=\\'0,1,2,3,4\\'.",
          "hmmCN_mu-Y=\\'-2,-0.45,0,0.3,0.51\\'.",
          "hmmCN_sigma-Y=\\'0.35,0.35,0.25,0.25,0.25\\'.",
          "diagonal-weight-Y=0.995.",
          "mapd-weight-Y=0.",
          "min-segment-size-Y=5.",
          "hmm-confidence-weight-Y=0.6.",
          "diagonal-weight=0.995.",
          "mapd-weight=0.",
          "min-segment-size=5.",
          "hmm-confidence-weight=0.6.",
          "diagonal-weight-X=0.995.",
          "mapd-weight-X=0.",
          "min-segment-size-X=5.",
          "hmm-confidence-weight-X=0.6.",
          "shrink=true ",
          " --analysis cn-cyto2-gender.cutoff=0.5 --analysis cn-segment --analysis lohCytoScan.lohCS_errorrate=0.05.lohCS_beta=0.001.lohCS_alpha=0.01.lohCS_separation=1000000.lohCS_nMinMarkers=10.lohCS_NoCallThreshold=0.05.lohCS_minGenomicSpan=1000000",
          " --qca-file ", file.path(object@analysis.path, object@qca, 
                fsep = .Platform$file.sep ),
          " --qcc-file ", file.path(object@analysis.path, object@qcc, 
                fsep = .Platform$file.sep ),
          " --snp-qc-snp-list ", file.path(object@analysis.path, object@snp, 
                fsep = .Platform$file.sep ),
          " --cdf-file ", file.path(object@analysis.path, object@cdf, 
                fsep = .Platform$file.sep ),
          " --chrX-probes ", file.path(object@analysis.path, object@chrX, 
                fsep = .Platform$file.sep ),
          " --chrY-probes ", file.path(object@analysis.path, object@chrY, 
                fsep = .Platform$file.sep ),
          " --annotation-file ", file.path(object@analysis.path, object@annot.db, 
                fsep = .Platform$file.sep ),
          " --reference-input ", file.path(object@analysis.path, object@refmodel, 
                fsep = .Platform$file.sep ),
          " --cel-files ", object@cel.path,
          " --out-dir ", object@output.path
        )
      }
    }

    if(object@os == "Windows") {
      if(object@level == "custom") {
        object@call <- paste0(tool, " ", object@params)
      } else {
        object@call <- paste0(tool, " -v 4",
          " --cyto2 false",
          " --doDualNormalization true",
          " --keep-intermediate-data false",
          " --run-geno-qc true", 
          " --snp-qc-use-contrast true",
          " --force true",
          " --adapter-type-normalization false",
          " --text-output true",
          " --cnchp-output false",
          " --cychp-output true",
          " --set-analysis-name cyhd",
          " --analysis loh-segment",
          " --analysis cn-neutral-loh",
          " --male-gender-ratio-cutoff 1.5",
          " --female-gender-ratio-cutoff 0.9",
          " --xx-cutoff 0.61",
          " --xx-cutoff-high 0.95",
          " --y-cutoff 0.58",
          " --local-gc-background-intensity-adjustment-method none",
          " --image-correction-intensity-adjustment-method none",
          " --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=6.wave-smooth=true",
          " --cn-calibrate-parameters calibrated-log2ratios.alpha-cn-calibrate=0.564278.alpha-X-cn-calibrate=0.619453.alpha-Y-cn-calibrate=0.494620.beta-cn-calibrate=1.beta-X-cn-calibrate=1.beta-Y-cn-calibrate=1",
          " --analysis genotype --analysis log2-ratio.gc-correction=false.median-autosome-median-normalization=true.median-smooth-marker-count=5 --log2ratio-adjustment-method log2ratio-adjustment-method-high-pass-filter.use=true",
          " --analysis allelic-difference-CytoScan.outlier-trim=3.0.step=20.window=100.point-count=128.bandwidth=0.25.cutoff=0.05.clean-threshold=0.35.symmetry=true ",
          " --analysis kernel-smooth.sigma_span=50",
          " --analysis cn-cyto2.",
          "hmmCN_state='0,1,2,3,4'.",
          "hmmCN_mu='-2,-0.45,0,0.3,0.51'.",
          "hmmCN_sigma='0.35,0.35,0.25,0.25,0.25'.",
          "hmmCN_state-X='0,1,2,3,4'.",
          "hmmCN_mu-X='-2,-0.47,0,0.33,0.53'.",
          "hmmCN_sigma-X='0.35,0.35,0.25,0.25,0.25'.",
          "hmmCN_state-Y='0,1,2,3,4'.",
          "hmmCN_mu-Y='-2,-0.45,0,0.3,0.51'.",
          "hmmCN_sigma-Y='0.35,0.35,0.25,0.25,0.25'.",
          "diagonal-weight-Y=0.995.",
          "mapd-weight-Y=0.",
          "min-segment-size-Y=5.",
          "hmm-confidence-weight-Y=0.6.",
          "diagonal-weight=0.995.",
          "mapd-weight=0.",
          "min-segment-size=5.",
          "hmm-confidence-weight=0.6.",
          "diagonal-weight-X=0.995.",
          "mapd-weight-X=0.",
          "min-segment-size-X=5.",
          "hmm-confidence-weight-X=0.6.",
          "shrink=true ",
          " --analysis cn-cyto2-gender.cutoff=0.5 --analysis cn-segment --analysis lohCytoScan.lohCS_errorrate=0.05.lohCS_beta=0.001.lohCS_alpha=0.01.lohCS_separation=1000000.lohCS_nMinMarkers=10.lohCS_NoCallThreshold=0.05.lohCS_minGenomicSpan=1000000",
          " --qca-file ", file.path(object@analysis.path, object@qca, 
              fsep = .Platform$file.sep ),
          " --qcc-file ", file.path(object@analysis.path, object@qcc, 
              fsep = .Platform$file.sep ),
          " --snp-qc-snp-list ", file.path(object@analysis.path, object@snp, 
              fsep = .Platform$file.sep ),
          " --cdf-file ", file.path(object@analysis.path, object@cdf, 
              fsep = .Platform$file.sep ),
          " --chrX-probes ", file.path(object@analysis.path, object@chrX, 
              fsep = .Platform$file.sep ),
          " --chrY-probes ", file.path(object@analysis.path, object@chrY, 
              fsep = .Platform$file.sep ),
          " --annotation-file ", file.path(object@analysis.path, object@annot.db, 
              fsep = .Platform$file.sep ),
          " --reference-input ", file.path(object@analysis.path, object@refmodel, 
              fsep = .Platform$file.sep ),
          " --cel-files ", object@cel.path,
          " --out-dir ", object@output.path
        )
      }
    }

    if(object@os == "MAC OS") {
      if(object@level == "custom") {
        object@call <- paste0(tool, " ", object@params)
      } else {
        object@call <- paste0(tool, " -v 4",
          " --cyto2 false",
        " --doDualNormalization true",
        " --keep-intermediate-data false",
        " --run-geno-qc true", 
        " --snp-qc-use-contrast true",
        " --force true",
        " --adapter-type-normalization false",
        " --text-output true",
        " --cnchp-output false",
        " --cychp-output true",
        " --set-analysis-name cyhd",
        " --analysis loh-segment",
        " --analysis cn-neutral-loh",
        " --male-gender-ratio-cutoff 1.5",
        " --female-gender-ratio-cutoff 0.9",
        " --xx-cutoff 0.61",
        " --xx-cutoff-high 0.95",
        " --y-cutoff 0.58",
        " --local-gc-background-intensity-adjustment-method none",
        " --image-correction-intensity-adjustment-method none",
        " --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=6.wave-smooth=true",
        " --cn-calibrate-parameters calibrated-log2ratios.alpha-cn-calibrate=0.564278.alpha-X-cn-calibrate=0.619453.alpha-Y-cn-calibrate=0.494620.beta-cn-calibrate=1.beta-X-cn-calibrate=1.beta-Y-cn-calibrate=1",
        " --analysis genotype --analysis log2-ratio.gc-correction=false.median-autosome-median-normalization=true.median-smooth-marker-count=5 --log2ratio-adjustment-method log2ratio-adjustment-method-high-pass-filter.use=true",
        " --analysis allelic-difference-CytoScan.outlier-trim=3.0.step=20.window=100.point-count=128.bandwidth=0.25.cutoff=0.05.clean-threshold=0.35.symmetry=true ",
        " --analysis kernel-smooth.sigma_span=50",
        " --analysis cn-cyto2.",
        "hmmCN_state=\\'0,1,2,3,4\\'.",
        "hmmCN_mu=\\'-2,-0.45,0,0.3,0.51\\'.",
        "hmmCN_sigma=\\'0.35,0.35,0.25,0.25,0.25\\'.",
        "hmmCN_state-X=\\'0,1,2,3,4\\'.",
        "hmmCN_mu-X=\\'-2,-0.47,0,0.33,0.53\\'.",
        "hmmCN_sigma-X=\\'0.35,0.35,0.25,0.25,0.25\\'.",
        "hmmCN_state-Y=\\'0,1,2,3,4\\'.",
        "hmmCN_mu-Y=\\'-2,-0.45,0,0.3,0.51\\'.",
        "hmmCN_sigma-Y=\\'0.35,0.35,0.25,0.25,0.25\\'.",
        "diagonal-weight-Y=0.995.",
        "mapd-weight-Y=0.",
        "min-segment-size-Y=5.",
        "hmm-confidence-weight-Y=0.6.",
        "diagonal-weight=0.995.",
        "mapd-weight=0.",
        "min-segment-size=5.",
        "hmm-confidence-weight=0.6.",
        "diagonal-weight-X=0.995.",
        "mapd-weight-X=0.",
        "min-segment-size-X=5.",
        "hmm-confidence-weight-X=0.6.",
        "shrink=true ",
          " --analysis cn-cyto2-gender.cutoff=0.5 --analysis cn-segment --analysis lohCytoScan.lohCS_errorrate=0.05.lohCS_beta=0.001.lohCS_alpha=0.01.lohCS_separation=1000000.lohCS_nMinMarkers=10.lohCS_NoCallThreshold=0.05.lohCS_minGenomicSpan=1000000",
          " --qca-file ", file.path(object@analysis.path, object@qca, 
                fsep = .Platform$file.sep ),
          " --qcc-file ", file.path(object@analysis.path, object@qcc, 
                fsep = .Platform$file.sep ),
          " --snp-qc-snp-list ", file.path(object@analysis.path, object@snp, 
                fsep = .Platform$file.sep ),
          " --cdf-file ", file.path(object@analysis.path, object@cdf, 
                fsep = .Platform$file.sep ),
          " --chrX-probes ", file.path(object@analysis.path, object@chrX, 
                fsep = .Platform$file.sep ),
          " --chrY-probes ", file.path(object@analysis.path, object@chrY, 
                fsep = .Platform$file.sep ),
          " --annotation-file ", file.path(object@analysis.path, object@annot.db, 
                fsep = .Platform$file.sep ),
          " --reference-input ", file.path(object@analysis.path, object@refmodel, 
                fsep = .Platform$file.sep ),
          " --cel-files ", object@cel.path,
          " --out-dir ", object@output.path
        )
      }
    }

	ScreenAffy("Calling APT")
    system(object@call)
	#message(object@call)
  }
)
